cygwin/CygwinMachine.c

```diff
@@ -0,0 +1,363 @@
+      CRT_fatalError("Cannot open " PROCSTATFILE);
+
+   this->boottime = -1;
+
+   while (true) {
+      char buffer[PROC_LINE_LENGTH + 1];
+      if (fgets(buffer, sizeof(buffer), statfile) == NULL)
+         break;
+      if (String_startsWith(buffer, "btime ") == false)
+         continue;
+      if (sscanf(buffer, "btime %lld\n", &this->boottime) == 1)
+         break;
+      CRT_fatalError("Failed to parse btime from " PROCSTATFILE);
+   }
+   fclose(statfile);
+
+   if (this->boottime == -1)
+      CRT_fatalError("No btime in " PROCSTATFILE);
+
+   CygwinMachine_updateCPUcount(this);
+
+   return super;
+}
+
+void Machine_delete(Machine* super) {
+   CygwinMachine* this = (CygwinMachine*) super;
+
+   Machine_done(super);
+   free(this->cpuData);
+   free(this);
+}
+
+static void CygwinMachine_scanMemoryInfo(CygwinMachine* this) {
+   Machine* host = &this->super;
+
+   // As for Cygwin 3.4.9,
+   // some of these fields are not present in /proc/meminfo
+
+   memory_t availableMem = 0;
+   memory_t freeMem = 0;
+   memory_t totalMem = 0;
+   memory_t buffersMem = 0;
+   memory_t cachedMem = 0;
+   memory_t sharedMem = 0;
+   memory_t swapTotalMem = 0;
+   memory_t swapCacheMem = 0;
+   memory_t swapFreeMem = 0;
+   memory_t sreclaimableMem = 0;
+
+   FILE* file = fopen(PROCMEMINFOFILE, "r");
+   if (!file)
+      CRT_fatalError("Cannot open " PROCMEMINFOFILE);
+
+   char buffer[128];
+   while (fgets(buffer, sizeof(buffer), file)) {
+
+      #define tryRead(label, variable)                                       \
+         if (String_startsWith(buffer, label)) {                             \
+            memory_t parsed_;                                                \
+            if (sscanf(buffer + strlen(label), "%llu kB", &parsed_) == 1) {  \
+               (variable) = parsed_;                                         \
+            }                                                                \
+            break;                                                           \
+         } else (void) 0 /* Require a ";" after the macro use. */
+
+      switch (buffer[0]) {
+      case 'M':
+         tryRead("MemAvailable:", availableMem);  // zero
+         tryRead("MemFree:", freeMem);
+         tryRead("MemTotal:", totalMem);
+         break;
+      case 'B':
+         tryRead("Buffers:", buffersMem);  // zero
+         break;
+      case 'C':
+         tryRead("Cached:", cachedMem);  // zero
+         break;
+      case 'S':
+         switch (buffer[1]) {
+         case 'h':
+            tryRead("Shmem:", sharedMem);  // zero
+            break;
+         case 'w':
+            tryRead("SwapTotal:", swapTotalMem);
+            tryRead("SwapCached:", swapCacheMem);  // zero
+            tryRead("SwapFree:", swapFreeMem);
+            break;
+         case 'R':
+            tryRead("SReclaimable:", sreclaimableMem);  // zero
+            break;
+         }
+         break;
+      }
+
+      #undef tryRead
+   }
+
+   fclose(file);
```

Comment:
Have a look at [`GlobalMemoryStatusEx`](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex). Further APIs should exist for even more details.

Reply:
is there any difference from directly reading cygwin procfs? i thought cygwin already maps win32 api very well.

Reply:
> Mostly performance and level of detail.
> 
> I'd prefer to avoid depending on both the cygwin and msys runtimes; and go fully native instead.

i see. but for now i will leave this work to someone who's more comfortable and familiar with windows apis.

about the performance actually it's not that bad, see https://github.com/htop-dev/htop/issues/1097#issuecomment-1737465996